### PR TITLE
refactor podUpdater interface

### LINKED ISSUE #182: Enhancement: `PodUpdater` need refactoring to provide more general interface
### What would you like to be added?

A new [PodUpdater](https://github.com/KusionStack/operating/blob/v0.3.0/pkg/controllers/collaset/synccontrol/update.go#L257)  interface to cover Pod update lifecycle, so that users can easily extend a new Pod upgrade type.

### Why is this needed?

[PodUpdater](https://github.com/KusionStack/operating/blob/v0.3.0/pkg/controllers/collaset/synccontrol/update.go#L257) currently is not adaptable since Replace upgrade type includes. We need to refactor it to represent the Pod update behavior better.

### PR DESCRIPTION
**Pull request by dbug-dk**
2. fix bug: replace pod by to-replace label, not check new pod status when pod upgrade policy is not replace update

#### 1. Does this PR affect any open issues?(Y/N) and add issue references 
fix #182 
- [ ] N
- [x] Y 

#### 2. What is the scope of this PR (e.g. component or file name):
collaset_controller reconcile:
1. syncPods
2. update

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [x] Other

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [x] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

#### 6. Release note

Please refer to [Release Notes Language Style Guide](https://kusionstack.io/docs/governance/release-policy/) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/controllers/collaset/synccontrol/update.go [hunk 1]
```diff
@@ -473,52 +609,159 @@ func (u *inPlaceOnlyPodUpdater) GetPodUpdateFinishStatus(_ *PodUpdateInfo) (fini
...
+	if err != nil {
+		collasetutils.AddOrUpdateCondition(resources.NewStatus,
+			appsv1alpha1.CollaSetScale, err, "UpdateFailed",
+			fmt.Sprintf("fail to update Context for updating: %s", err))
+		return updating, recordedRequeueAfter, err
+	} else {
+		collasetutils.AddOrUpdateCondition(resources.NewStatus,
+			appsv1alpha1.CollaSetScale, nil, "UpdateFailed", "")
+	}
+	return updating, nil, nil
```
Comment (wu8685) [inline pkg/controllers/collaset/synccontrol/update.go:667]: missing `recordedRequeueAfter` to return

File: pkg/controllers/collaset/synccontrol/sync_control.go [hunk 2]
```diff
@@ -663,124 +668,17 @@ func (r *RealSyncControl) Update(
...
+	updating, recordedRequeueAfter, err := updater.PrepareAndFilterPodUpdate(podToUpdate, resources, ownedIDs, podCh)
+	if err != nil {
+		return updating, recordedRequeueAfter, err
 	}
 
 	// 6. update Pod
 	succCount, err := controllerutils.SlowStartBatch(len(podCh), controllerutils.SlowStartInitialBatchSize, false, func(_ int, _ error) error {
 		podInfo := <-podCh
-		if !analysedPod.Has(podInfo.Name) {
+		if !podInfo.analysed {
```
Comment (wu8685) [inline pkg/controllers/collaset/synccontrol/sync_control.go:681]: Do we still need this `analysed` here?

File: pkg/controllers/collaset/synccontrol/update.go [hunk 3]
```diff
@@ -253,23 +265,26 @@ func (o orderByDefault) Less(i, j int) bool {
 }
 
 type PodUpdater interface {
+	PrepareAndFilterPodUpdate(podToUpdate []*PodUpdateInfo, resources *collasetutils.RelatedResources,
```
Comment (wu8685) [inline pkg/controllers/collaset/synccontrol/update.go:268]: Two points I find need to discuss:
1. I think `PrepareAndFilterPodUpdate` could be split into two methods: `FilterPodToUpdate` and `PreparePodUpdate`
2. `Recreate` and `InPlaceIfPossible` have same implementation. I think they can share the same top Updater and be dispatched by another Interface.

File: pkg/controllers/collaset/synccontrol/update.go [hunk 4]
```diff
@@ -473,52 +609,159 @@ func (u *inPlaceOnlyPodUpdater) GetPodUpdateFinishStatus(_ *PodUpdateInfo) (fini
...
+	return recreatePod(u.collaSet, podInfo, u.podControl, u.recorder)
+}
+
 func (u *recreatePodUpdater) GetPodUpdateFinishStatus(podInfo *PodUpdateInfo) (finished bool, msg string, err error) {
 	// Recreate policy alway treat Pod as update finished
 	return podInfo.IsUpdatedRevision && !podInfo.PodDecorationChanged, "", nil
 }
 
+func (u *recreatePodUpdater) FinishUpdatePod(podInfo *PodUpdateInfo) error {
+	//u.recorder.Eventf().V(1).Info("try to finish update PodOpsLifecycle for Pod", "pod", commonutils.ObjectKeyString(podInfo.Pod))
```
Comment (wu8685) [inline pkg/controllers/collaset/synccontrol/update.go:684]: remove this comment

File: pkg/controllers/collaset/synccontrol/update.go [hunk 5]
```diff
@@ -253,23 +265,26 @@ func (o orderByDefault) Less(i, j int) bool {
 }
 
 type PodUpdater interface {
+	PrepareAndFilterPodUpdate(podToUpdate []*PodUpdateInfo, resources *collasetutils.RelatedResources,
+		ownedIDs map[int]*appsv1alpha1.ContextDetail, podCh chan *PodUpdateInfo) (bool, *time.Duration, error)
 	FulfillPodUpdatedInfo(revision *appsv1.ControllerRevision, podUpdateInfo *PodUpdateInfo) error
```
Comment (wu8685) [inline pkg/controllers/collaset/synccontrol/update.go:270]: `FulfillPodUpdatedInfo` has been called in two places. Can we refactor it?

**Comment by ColdsteelRail** [comment]
move to pr #185

**Review by ColdsteelRail — APPROVED** [review]
lgtm

**Comment by wu8685** [comment]
Shouldn't we return `false` [here](https://github.com/dbug-dk/operating/blob/updater-refactor/pkg/controllers/collaset/synccontrol/update.go#L755)?

**Comment by dbug-dk** [comment]
> Shouldn't we return `false` [here](https://github.com/dbug-dk/operating/blob/updater-refactor/pkg/controllers/collaset/synccontrol/update.go#L755)?

yes, I'll change this.

**Review by wu8685 — APPROVED** [review]
lgtm

**Pull request merged**